File: src/interception/state.py
import math
import random

from .collision import swept_hit
from .fleet import Fleet
from .obstacle import Obstacle
from .params import ScenarioParams
from .target import Target


class State:
    """
    The world: extent in metres, a seeded RNG, and every entity in it.

    Pygame-free by construction, so the whole model can run headless in CI
    without an SDL surface.
    """

    def __init__(self, seed, scenario=None):
        self.scenario = scenario or ScenarioParams()
        self.width = self.scenario.world_width_m
        self.height = self.scenario.world_height_m

        # A concrete seed is required. Callers that want an arbitrary world
        # draw a seed first and record it, so every run stays reproducible.
        self.seed = int(seed)
        self.rng = random.Random(self.seed)

        self.obstacles = []
        self.targets = []
        self.agents = []
        self.actors = []
        self.intercepts = 0
        self.min_miss_distance_m = math.inf

        self._build_world()

    def _build_world(self):
        cfg = self.scenario
        n_obstacles = self.rng.randint(cfg.min_obstacles, cfg.max_obstacles)
        self.obstacles = [Obstacle(self) for _ in range(n_obstacles)]
        self.targets = [Target(self, cfg.target) for _ in range(cfg.num_targets)]
        self.fleet = Fleet(cfg.num_agents, self, cfg.interceptor, cfg.guidance)
        self.agents = list(self.fleet.agents)
        # Obstacles are static, so they are not actors and never step.
        self.actors = self.agents + self.targets
        self.intercepts = 0
        self.min_miss_distance_m = math.inf

    def reset(self, seed=None):
        """
        Rebuild the world from a known seed.

        Passing ``None`` reuses the seed this world was built with rather
        than reseeding from OS entropy, so ``reset()`` is repeatable.
        """
        if seed is not None:
            self.seed = int(seed)
        self.rng = random.Random(self.seed)
        self._build_world()

    def update(self, dt):
        for actor in self.actors:
            actor.step(dt)
        self._resolve_intercepts()

    def _resolve_intercepts(self):
        """
        Swept-sphere test over every agent/target pair for this step.

        Removals are deferred until after the scan: mutating ``targets`` or
        ``actors`` mid-iteration silently skips entries.
        """
        doomed = {}
        for agent in self.agents:
            for target in self.targets:
                combined_radius = agent.hit_radius_m + target.hit_radius_m
                hit, miss = swept_hit(
                    agent.prev_pos,
                    agent.pos,
                    target.prev_pos,
                    target.pos,
                    combined_radius,
                )
                if miss < self.min_miss_distance_m:
                    self.min_miss_distance_m = miss
                if hit:
                    doomed[id(target)] = target

        if not doomed:
            return

        self.targets = [t for t in self.targets if id(t) not in doomed]
        self.actors = [a for a in self.actors if id(a) not in doomed]
        self.intercepts += len(doomed)

File: core/src/state.cpp
#include "interception/state.hpp"

#include <cmath>
#include <limits>
#include <utility>

#include "interception/collision.hpp"

namespace interception {

State::State(std::uint64_t seed, ScenarioParams scenario)
    : scenario_(std::move(scenario)), seed_(seed), rng_(seed) {
    buildWorld();
}

void State::reset() { reset(seed_); }

void State::reset(std::uint64_t seed) {
    seed_ = seed;
    rng_.reseed(seed_);
    buildWorld();
}

void State::buildWorld() {
    obstacles_.clear();
    targets_.clear();
    agents_.clear();

    // Construction order is part of the seed contract: obstacles, then
    // targets, then agents. Each pulls a fixed number of draws, so
    // reordering these loops changes every world.
    const int obstacleCount = rng_.randint(scenario_.minObstacles, scenario_.maxObstacles);
    obstacles_.reserve(static_cast<std::size_t>(obstacleCount));
    for (int i = 0; i < obstacleCount; ++i) {
        obstacles_.push_back(makeObstacle(rng_, widthM(), heightM()));
    }

    targets_.reserve(static_cast<std::size_t>(scenario_.numTargets));
    for (int i = 0; i < scenario_.numTargets; ++i) {
        targets_.emplace_back(*this, scenario_.target, rng_);
    }

    agents_.reserve(static_cast<std::size_t>(scenario_.numAgents));
    for (int i = 0; i < scenario_.numAgents; ++i) {
        agents_.emplace_back(*this, scenario_.interceptor, scenario_.guidance, rng_);
    }

    intercepts_ = 0;
    minMissDistanceM_ = std::numeric_limits<double>::infinity();
}

void State::update(double dt) {
    // Agents before targets, matching the reference implementation's
    // actor ordering. Obstacles are static and never step.
    for (Agent& agent : agents_) {
        agent.step(dt);
    }
    for (Target& target : targets_) {
        target.step(dt);
    }
    resolveIntercepts();
}

void State::resolveIntercepts() {
    if (targets_.empty()) {
        return;
    }

    std::vector<bool> doomed(targets_.size(), false);
    int doomedCount = 0;

    for (const Agent& agent : agents_) {
        for (std::size_t i = 0; i < targets_.size(); ++i) {
            const Target& target = targets_[i];
            const double combinedRadius = agent.hitRadiusM + target.hitRadiusM;
            const SweptHitResult result = sweptHit(agent.prevPos, agent.pos,
                                                   target.prevPos, target.pos,
                                                   combinedRadius);
            if (result.missDistanceM < minMissDistanceM_) {
                minMissDistanceM_ = result.missDistanceM;
            }
            if (result.hit && !doomed[i]) {
                doomed[i] = true;
                ++doomedCount;
            }
        }
    }

    if (doomedCount == 0) {
        return;
    }

    std::vector<Target> survivors;
    survivors.reserve(targets_.size() - static_cast<std::size_t>(doomedCount));
    for (std::size_t i = 0; i < targets_.size(); ++i) {
        if (!doomed[i]) {
            survivors.push_back(std::move(targets_[i]));
        }
    }
    targets_ = std::move(survivors);
    intercepts_ += doomedCount;
}

bool State::hasMinMissDistance() const { return std::isfinite(minMissDistanceM_); }

double State::totalDeltaVMps() const {
    double total = 0.0;
    for (const Agent& agent : agents_) {
        total += agent.deltaVMps;
    }
    return total;
}

}  // namespace interception
